# Bài tập 5

Bùi Chí Dũng - 1712364

## Câu 1 (1 điểm)

Giải bất phương trình ta sẽ nhận được giá trị N >= 45 để expected E_in >= 0.08. Vì vậy, theo đáp án, ít nhất N phải đạt 100. Chọn [c] 100, nếu N = 25 thì không thể đạt E_in ~0.08 được.

## Câu 2 (1 điểm)

Ta có: $\tilde{w}_0 + \tilde{w}_1 x_1^2 + \tilde{w}_2 x_2^2 = 0$
- $x_1$ nhỏ -> dương, $x_1$ lớn -> âm
- $x_2$ nhỏ -> âm dương âm, $x_2$ lớn -> dương
Vì y = x và y = -x là các đường liên tục, nên $\tilde w_1 = - \tilde w_2$
.Mặt khác, ban đầu là dương nên ta biết $\tilde w_1$ dương, và $x_1$ lớn thì âm nên $\tilde w_1$ âm.
Vì vậy mà $ \tilde w_1$ < 0, $ \tilde w_2$ > 0. Chọn [e] $\tilde{w}_1 < 0 ,  \tilde{w}_2 > 0$


## Câu 3 (1 điểm)

Cho $\Phi_4$ (15 outputs) với 2 inputs ($\mathbf x \in \mathbb R^2 $)
Ta có: 
$d_{VC} = d + 1 = 3$ (2D perceptron). 
Mặt khác, $d_{VC} = \tilde d + 1$, trong đó $\tilde d + 1 = 15$ (số lượng outputs của $\Phi$).
Chọn [c] 15



## Câu 4 (1 điểm)

Đạo hàm đa thức E(u,v) theo biến u ta được kết quả là [e] $2(e^v + 2ve^{-u})(ue^v - 2ve^{-u})$

In [1]:
import numpy as np

def E(x):
    u = x[0]
    v = x[1]
    return (u*np.exp(v) - 2*v*np.exp(-u))**2

def grad_E(x):
    u = x[0]
    v = x[1]
    return np.array([2*(u*np.exp(v) - 2*v*np.exp(-u))*(np.exp(v) + 2*v*np.exp(-u)),
                     2*(u*np.exp(v) - 2*v*np.exp(-u))*(u*np.exp(v) - 2*np.exp(-u))])

def gradient_descent(E, grad_E, x0, tol=1e-14, eta=0.1):
    x = x0.copy()
    iters = 0
    err = E(x)
    us = [x[0]]
    vs = [x[1]]
    while err > tol:
        grad = grad_E(x)
        x = x - eta*grad
        us.append(x[0])
        vs.append(x[1])
        err = E(x)
        iters += 1
        if iters > 100:
            print("Không hội tụ")
            return
    return (x, err, iters, us, vs)
x0 = np.array([1, 1])
final_uv, err, iters, us, vs = gradient_descent(E, grad_E, x0)
print("Iterations:", iters)
print("Final u,v:", final_uv)

Iterations: 10
Final u,v: [0.04473629 0.02395871]


## Câu 5 (1 điểm)

Từ kết quả trên ta chọn được câu [d] 10 interations.

## Câu 6 (1 điểm)

Từ kết quả trên ta chọn được câu [e] (0.045, 0.024)

## Câu 7 (1 điểm)

In [2]:
import numpy as np

def E(u, v):
    return (u*np.exp(v) - 2*v*np.exp(-u))**2

def dE_u(u, v):
    return 2*(u*np.exp(v) - 2*v*np.exp(-u))*(np.exp(v) + 2*v*np.exp(-u))

def dE_v(u, v):
    return 2*(u*np.exp(v) - 2*v*np.exp(-u))*(u*np.exp(v) - 2*np.exp(-u))
def coordinate_descent(E, dE_u, dE_v, u0, v0, tol =1e-14, eta = 0.1):
    u, v = u0, v0
    us = [u]
    uv = [v]
    for it in range(15):
        u = u - eta*dE_u(u,v)
        us.append(u)
        vs.append(v)
        v = v - eta*dE_v(u,v)
        us.append(u)
        vs.append(v)
    return u, v, E(u, v), 15, us, vs
final_u, final_v, err, iters, us, vs = coordinate_descent(E, dE_u, dE_v, 1.0, 1.0)
print("Final coordinate:", final_u, final_v)
print("Final error:", err)
print("Iterations:", iters)

Final coordinate: 6.29707589930517 -2.852306954077811
Final error: 0.13981379199615315
Iterations: 15


Theo kết quả trên thì the Error E(u,v) gần bằng 0.14 ~ 0.1. Vậy chọn [a] $10^{-1}$

## Câu 8 (1 điểm)

Dự đoán giá trị sẽ thay đổi xung quanh 0.1, nên em chọn [d] 0.100

## Câu 9 (1 điểm)

Dự đoán kết quả thay đổi xung quanh giá trị 350, nên em chọn [a] 350

## Câu 10 (1 điểm)

Theo PLA ta có vòng lặp để cập nhật như sau:
$w = w + y_nx_n$ đối với những điểm được phân loại sai ($y_n \neq w^T x_n$) và w không đổi với những điểm phân loại đúng.
Đối với SGD ta có vòng lặp cập nhật sau:
$w = w - \alpha \nabla e_n(w) $, đặt $\alpha = 1$. Chọn $e_n$ để $\nabla e_n \approx - y_n x_n $ cho những điểm được phân loại sai và $\nabla e_n \approx 0 $ đối với những điểm phân loại đúng.



- [a] Gradient:  $- y_n x_n e^{-y_n w^T x_n}$ . Với những điểm phân loại đúng ta được $-y_nx_n$ (n nhỏ), với những điểm phân loại sai ta được $-y_nx_n$ (với n lớn)

- [b] Gradient: $-y_nx_n$. Sai vì kết quả vẫn là $-y_nx_n$ dù cho điểm được phân loại đúng hay sai

- [c] Gradient:  $-2(y_n - w^T x_n)x_n$. Sai, vì sự khác nhau giữa $y_n$ và $w^T x_n$ 

- [d] Gradient: $$ \frac{-y_n x_n}{1 + e^{y_n w^T x_n}} $$
Với những điểm phân loại sai ta có: $-y_nx_n$ (n nhỏ)
Với những điểm phân loại đúng ta có $-y_nx_n$

- [e] Với những điểm phân loại đúng ta có gradient = 0, với những điểm phân loại sai ta có gradient = $-y_nx_n$. Kết quả này giống như thuật toán PLA. 


Vậy chọn [e] $e_n(w) = - min(0, y_nw^Tx_n)$